File: dimdwarf-core/src/main/java/net/orfjackal/dimdwarf/net/SimpleSgsProtocolDecoder.java
// Copyright © 2008-2010 Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://dimdwarf.sourceforge.net/LICENSE

package net.orfjackal.dimdwarf.net;

import org.apache.mina.core.buffer.IoBuffer;
import org.apache.mina.core.session.IoSession;
import org.apache.mina.filter.codec.*;

public class SimpleSgsProtocolDecoder extends CumulativeProtocolDecoder {

    // TODO: ensure compatibility with com.sun.sgs.impl.sharedutil.MessageBuffer

    protected boolean doDecode(IoSession session, IoBuffer in, ProtocolDecoderOutput out) throws Exception {

        // "doDecode() can't return true when buffer is not consumed."
        for (int i = 0; i < in.limit(); i++) {
            in.get();
        }

        out.write(new LoginRequest());
        return true;
    }
}

File: dimdwarf-core/src/main/scala/net/orfjackal/dimdwarf/net/SimpleSgsProtocolDecoder.scala
package net.orfjackal.dimdwarf.net

import org.apache.mina.core.buffer.IoBuffer
import com.sun.sgs.protocol.simple.SimpleSgsProtocol
import org.apache.mina.filter.codec._
import org.apache.mina.core.session.IoSession
import javax.annotation.concurrent.Immutable
import java.nio.charset.Charset
import com.sun.sgs.impl.sharedutil.MessageBuffer

@Immutable
class SimpleSgsProtocolDecoder extends CumulativeProtocolDecoder {
  private val stringCharset = Charset.forName("UTF-8")

  protected def doDecode(session: IoSession, in: IoBuffer, out: ProtocolDecoderOutput): Boolean = {
    if (in.prefixedDataAvailable(2, SimpleSgsProtocol.MAX_PAYLOAD_LENGTH)) {
      val payloadLength = in.getUnsignedShort()
      val op = in.get()

      val message = op match {
        case SimpleSgsProtocol.LOGIN_REQUEST =>
          val version = in.get()
          require(version == SimpleSgsProtocol.VERSION, "incompatible version: " + version)
          val username = readString(in)
          val password = readString(in)
          LoginRequest(username, password)
      }

      out.write(message)
      true
    } else {
      false
    }
  }

  private def readString(in: IoBuffer): String = {
    val length = in.getUnsignedShort()
    val bytes = new Array[Byte](length)
    in.get(bytes)

    val buf = new MessageBuffer(2 + length).
            putShort(length).
            putBytes(bytes)
    buf.rewind()
    buf.getString()
  }
}
